### DISCUSSION TITLE
Optimizing scene, selecting geometries and changing materials

### DISCUSSION THREAD

**Discussion by peterbako** [openscenegraph/OpenSceneGraph#1215] [Q&A] [answered]
Hello,

I need an advice. I am developing a desktop application for simulation of CNC machines using OSG 3.1.3.
Since some time I started to face some performance issues at rendering. I found out that the reason is probably using too many geometries (osg::Geometry objects), which makes the rendering slow at some cases. 
I splitted the scene in the tree so, that each face of a solid is represented by one osg::Geometry (a derived class in my case). 
each osg::Geode contains a list of faces (osg::Geometry) and the geode represents a solid
In one case i get 2,2 million triangles and 150000 faces (geometries) 1280 solids and 1276 nodes. By this scene i get about 1-2 FPS which is terribly slow.

I need to support different material per face, face highlighting (using material), transparency, clipping planes, therefore this was my base decision to define the scene in this way, but now I came to the limits.

I decided to make geometry optimization so, that I group the geometries together by color, so that the geometries with same color share the same list of vertices, normals, colors and the triangles are defined with lists of DrawElementUInt(GL_TRIANGLES) per face. Additionally I am storing the face information in UserData of osg::Geometry in an own datastructure. I wanted to derive from DrawElementUInt and store there but my additional data were not serialized, but this is a different topic.

This optimization brought me significant performance boost, I get about 15-25 FPS by the same scene, and that is fine.
The drawback is that if I need to change a color per face, I need to rebuild the structure - create a new geometry with the specific data per face and remove it from the old one which brings me significant problems in the current situation (thread synchronization-> memory corruption and crashes).

Do you have an other idea to make this optimization in a better way?
Thank you very much for each comment!

Best regards
Peter

**Comment by vsg-dev** [comment, marked as answer]
The graphics hardware/OepnGL just supports changing colour per primitive set or per vertex, so per face material changes you'll need to either ease separate PrimitiveSet per face or duplicate vertices in order to use per vertex colours.

Potentially you could subclass from osg::Drawable and implement the GL calls tuned for your usage case and for the subclass to have the public variables/member variables you need.

Perhaps if you have hardware that supports it you could try out RTX mesh shaders. For this you'd need to the MeshShaders branch of the OSG.

Longer term you might also want to consider the VulkanSceneGraph, both Vulkan and the VSG itself handle scenes where meshes are broken into small chunks far better than OpenGL & OSG as the CPU overhead is far far lower.  They also have support for cross vendor mesh shaders.

**Reply by peterbako** [reply]
Hello,
Thanks for the answer and sorry for the late response.
I am not only setting colors, I am also setting materials, and this is only possible per geometry, not per primitiveSet.

I am not so skilled in openGL or shaders, so I don't want to go this direction.

Finally, I solved the optimization and the crashes were the result of incorrect geometry creations.
VSG would be a topic for years, I read about it but currently it doesn't come into the possible solutions.
Thanks for your efforts and advices.

Best regards
Peter
